File: src/main/java/no/nav/helse/flex/infrastructure/kafka/JfrTopologies.java
package no.nav.helse.flex.infrastructure.kafka;

import com.google.gson.Gson;
import io.confluent.kafka.streams.serdes.avro.GenericAvroSerde;
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.EventEnricherTransformerSupplier;
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.GenerellOperationsTransformerSupplier;
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.JournalOperationsTransformerSupplier;
import no.nav.helse.flex.operations.Feilregistrer;
import no.nav.helse.flex.operations.SkjemaMetadata;
import no.nav.helse.flex.Environment;
import no.nav.helse.flex.infrastructure.metrics.Metrics;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.common.serialization.Serde;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.Topology;
import org.apache.kafka.streams.kstream.Consumed;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.Produced;
import org.apache.kafka.streams.state.KeyValueStore;
import org.apache.kafka.streams.state.StoreBuilder;
import org.apache.kafka.streams.state.Stores;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class JfrTopologies {
    private static final Logger log = LoggerFactory.getLogger(JfrTopologies.class);
    private static final List<String> eventType = Arrays.asList("MidlertidigJournalført", "Mottatt", "JournalpostMottatt");

    private final String inputTopic;
    private final String manuellTopic;
    private final String ENRICHER_OPERATION_STORE = "enrichoperations";
    private final String GENERELL_OPERATION_STORE = "generelloperations";
    private final String JOURNALFOERING_OPERATION_STORE = "journalfoeringoperations";
    private final Serde<KafkaEvent> kafkaEventSerde = Serdes.serdeFrom(new JfrKafkaSerializer<>(), new JfrKafkaDeserializer<>(KafkaEvent.class));
    private final Serde<EnrichedKafkaEvent> enhancedKafkaEventSerde = Serdes.serdeFrom(new JfrKafkaSerializer<>(), new JfrKafkaDeserializer<>(EnrichedKafkaEvent.class));

    private final StoreBuilder<KeyValueStore<String, EnrichedKafkaEvent>> eventEnricherSupplier;
    private final StoreBuilder<KeyValueStore<String, EnrichedKafkaEvent>> generellOperationSupplier;
    private final StoreBuilder<KeyValueStore<String, EnrichedKafkaEvent>> journalOperationSupplier;
    private final SkjemaMetadata skjemaMetadata;
    private final EventEnricherTransformerSupplier eventEnricherTransformerSupplier;
    private final GenerellOperationsTransformerSupplier generellOperationsTransformerSupplier;
    private final JournalOperationsTransformerSupplier journalOperationsTransformerSupplier;
    private final Feilregistrer feilregistrer;

    public JfrTopologies(final String inputTopic, final String manuellTopic) throws Exception {
        this.inputTopic = inputTopic;
        this.manuellTopic = manuellTopic;
        this.skjemaMetadata = new SkjemaMetadata();
        this.eventEnricherTransformerSupplier = new EventEnricherTransformerSupplier(ENRICHER_OPERATION_STORE);
        this.generellOperationsTransformerSupplier = new GenerellOperationsTransformerSupplier(GENERELL_OPERATION_STORE);
        this.journalOperationsTransformerSupplier = new JournalOperationsTransformerSupplier(JOURNALFOERING_OPERATION_STORE);
        this.feilregistrer = new Feilregistrer();
        this.eventEnricherSupplier = Stores.keyValueStoreBuilder(
                Stores.inMemoryKeyValueStore(ENRICHER_OPERATION_STORE),
                Serdes.String(),
                enhancedKafkaEventSerde);
        this.generellOperationSupplier = Stores.keyValueStoreBuilder(
                Stores.inMemoryKeyValueStore(GENERELL_OPERATION_STORE),
                Serdes.String(),
                enhancedKafkaEventSerde);
        this.journalOperationSupplier = Stores.keyValueStoreBuilder(
                Stores.inMemoryKeyValueStore(JOURNALFOERING_OPERATION_STORE),
                Serdes.String(),
                enhancedKafkaEventSerde);
    }

    public Topology getJfrTopologi() {
        final StreamsBuilder streamsBuilder = new StreamsBuilder();
        final Map<String, String> serdeConfig = Environment.getKafkaSerdeConfig();
        final Serde<GenericRecord> valueGenericAvroSerde = new GenericAvroSerde();
        valueGenericAvroSerde.configure(serdeConfig, false); // `false` for record values
        streamsBuilder.addStateStore(eventEnricherSupplier);
        streamsBuilder.addStateStore(generellOperationSupplier);
        streamsBuilder.addStateStore(journalOperationSupplier);

        final KStream<String, GenericRecord> aapenDokStream = streamsBuilder.stream(inputTopic, Consumed.with(Serdes.String(), valueGenericAvroSerde));
        final KStream<String, GenericRecord> filteredEvents = filterGenerelleEvent(aapenDokStream);
        final KStream<String, KafkaEvent> kafkaEventKStream = convertGenericRecordToKafkaEvent(filteredEvents);
        final KStream<String, EnrichedKafkaEvent> enrichedEventsStream = enrichKafkaEvent(kafkaEventKStream);

        final KStream<String, EnrichedKafkaEvent> filterJournalpostToAuto = filterJournalpostToAuto(enrichedEventsStream);
        final KStream<String, EnrichedKafkaEvent> processedGenerellStream = generellOperations(filterJournalpostToAuto);
        journalfoerJournalpost(processedGenerellStream);
        return streamsBuilder.build();
    }

    private KStream<String, GenericRecord> filterGenerelleEvent(final KStream<String, GenericRecord> inputStream) {
        return inputStream
                .filter((key, genericRecord) ->
                        genericRecord.get("temaNytt").toString().equals("SYK")
                        && eventType.contains(genericRecord.get("hendelsesType").toString()));
    }

    private KStream<String, KafkaEvent> convertGenericRecordToKafkaEvent(final KStream<String, GenericRecord> genericRecordString) {
        final KStream<String, KafkaEvent> kafkaEventKStream = genericRecordString.mapValues((genericRecord -> new Gson().fromJson(genericRecord.toString(), KafkaEvent.class)));
        return kafkaEventKStream;
    }

    private KStream<String, EnrichedKafkaEvent> enrichKafkaEvent(final KStream<String, KafkaEvent> kafkaEventKStream) {
        final KStream<String, EnrichedKafkaEvent> enrichedKafkaEvent = kafkaEventKStream.transform(eventEnricherTransformerSupplier, ENRICHER_OPERATION_STORE);
        final KStream<String, EnrichedKafkaEvent> toManuell = enrichedKafkaEvent.filter(((s, jfrEnhancedKafkaEvent) -> jfrEnhancedKafkaEvent.isToManuell() && !jfrEnhancedKafkaEvent.isToIgnore()));
        sendToJfrManuellOppretter(toManuell);
        return enrichedKafkaEvent.filter(((s, jfrEnhancedKafkaEvent) -> !jfrEnhancedKafkaEvent.isToManuell() && !jfrEnhancedKafkaEvent.isToIgnore()));
    }

    private KStream<String, EnrichedKafkaEvent> filterJournalpostToAuto(final KStream<String, EnrichedKafkaEvent> enrichedKafkaEventStream) {
        final KStream<String, EnrichedKafkaEvent> toAuto = enrichedKafkaEventStream.filter(((s, jfrEnhancedKafkaEvent) -> skjemaMetadata.inAutoList(jfrEnhancedKafkaEvent.getTema(), jfrEnhancedKafkaEvent.getSkjema())))
                .peek((k, enrichedKafkaEvent) -> logWithCorrelationId(enrichedKafkaEvent, "Forsøker automatisk behandling på journalpost {} med tema {} og skjema {}",
                        enrichedKafkaEvent.getJournalpostId(), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getSkjema()));
        final KStream<String, EnrichedKafkaEvent> toManuell = enrichedKafkaEventStream.filterNot(((s, jfrEnhancedKafkaEvent) -> skjemaMetadata.inAutoList(jfrEnhancedKafkaEvent.getTema(), jfrEnhancedKafkaEvent.getSkjema())));
        sendToJfrManuellOppretter(toManuell);
        return toAuto;
    }

    private KStream<String, EnrichedKafkaEvent> generellOperations(final KStream<String, EnrichedKafkaEvent> enrichedKafkaEventStream) {
        final KStream<String, EnrichedKafkaEvent> afterGenerellOperationStream = enrichedKafkaEventStream.transform(generellOperationsTransformerSupplier, GENERELL_OPERATION_STORE);
        final KStream<String, EnrichedKafkaEvent> toManuell = afterGenerellOperationStream.filter((k, enrichedKafkaEvent) -> enrichedKafkaEvent.isToManuell());
        sendToJfrManuellOppretter(toManuell);
        return afterGenerellOperationStream.filterNot((k, enrichedKafkaEvent) -> enrichedKafkaEvent.isToManuell());
    }

    private void journalfoerJournalpost(final KStream<String, EnrichedKafkaEvent> oppgaveFilteredStream) {
        final KStream<String, EnrichedKafkaEvent> journalpostAfterJournalfoering = oppgaveFilteredStream.transform(journalOperationsTransformerSupplier, JOURNALFOERING_OPERATION_STORE);
        final KStream<String, EnrichedKafkaEvent> toManuell = journalpostAfterJournalfoering.filter((k, enrichedKafkaEvent) -> enrichedKafkaEvent.isToManuell());
        final KStream<String, EnrichedKafkaEvent> doneProcessing = journalpostAfterJournalfoering.filterNot((k, enrichedKafkaEvent) -> enrichedKafkaEvent.isToManuell());
        doneProcessing.peek((s, enrichedKafkaEvent) -> Metrics.incJfrAutoProcess(enrichedKafkaEvent));
        sendToJfrManuellOppretter(toManuell);
    }

    private void sendToJfrManuellOppretter(final KStream<String, EnrichedKafkaEvent> manuelle) {
        manuelle.foreach(feilregistrer::feilregistrerOppgave);
        manuelle.peek((s, enrichedKafkaEvent) -> Metrics.incJfrManuallProcess(enrichedKafkaEvent, skjemaMetadata.inAutoList(enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getSkjema())));
        manuelle.peek((k, enrichKafkaEvent) -> logWithCorrelationId(enrichKafkaEvent, "Journalposten: {} sendes til manuell-oppretter", enrichKafkaEvent.getJournalpostId()));
                // TODO: Slå på når topic og app er opprettet
                // .to(manuellTopic, Produced.with(Serdes.String(), enhancedKafkaEventSerde));
    }

    private void logWithCorrelationId(EnrichedKafkaEvent enrichedKafkaEvent, String s, String... args) {
        MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
        log.info(s, (Object[]) args);
        MDC.clear();
    }

}

File: src/main/kotlin/no/nav/helse/flex/infrastructure/kafka/JfrTopologies.kt
package no.nav.helse.flex.infrastructure.kafka

import com.fasterxml.jackson.module.kotlin.readValue
import io.confluent.kafka.streams.serdes.avro.GenericAvroSerde
import no.nav.helse.flex.Environment.kafkaSerdeConfig
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.EventEnricherTransformerSupplier
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.GenerellOperationsTransformerSupplier
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.JournalOperationsTransformerSupplier
import no.nav.helse.flex.infrastructure.kafka.transformerSupplier.OppgaveOperationsTransformerSupplier
import no.nav.helse.flex.infrastructure.metrics.Metrics.incJfrAutoProcess
import no.nav.helse.flex.infrastructure.metrics.Metrics.incJfrManuallProcess
import no.nav.helse.flex.objectMapper
import no.nav.helse.flex.operations.Feilregistrer
import no.nav.helse.flex.operations.SkjemaMetadata
import org.apache.avro.generic.GenericRecord
import org.apache.kafka.common.serialization.Serde
import org.apache.kafka.common.serialization.Serdes
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.Topology
import org.apache.kafka.streams.kstream.Consumed
import org.apache.kafka.streams.kstream.KStream
import org.apache.kafka.streams.state.Stores
import org.slf4j.LoggerFactory
import org.slf4j.MDC

class JfrTopologies(
    private val inputTopic: String,
    private val feilregistrer: Feilregistrer = Feilregistrer(),
    private val eventEnricherTransformerSupplier: EventEnricherTransformerSupplier = EventEnricherTransformerSupplier(
        ENRICHER_OPERATION_STORE
    ),
    private val generellOperationsTransformerSupplier: GenerellOperationsTransformerSupplier = GenerellOperationsTransformerSupplier(
        GENERELL_OPERATION_STORE
    ),
    private val journalOperationsTransformerSupplier: JournalOperationsTransformerSupplier = JournalOperationsTransformerSupplier(
        JOURNALFOERING_OPERATION_STORE
    ),
    private val oppgaveOperationsTransformerSupplier: OppgaveOperationsTransformerSupplier = OppgaveOperationsTransformerSupplier(
        OPPGAVE_OPERATION_STORE
    )
) {

    private val enhancedKafkaEventSerde = Serdes.serdeFrom(
        JfrKafkaSerializer(),
        JfrKafkaDeserializer(
            EnrichedKafkaEvent::class.java
        )
    )
    private val eventEnricherSupplier = Stores.keyValueStoreBuilder(
        Stores.inMemoryKeyValueStore(ENRICHER_OPERATION_STORE),
        Serdes.String(),
        enhancedKafkaEventSerde
    )
    private val generellOperationSupplier = Stores.keyValueStoreBuilder(
        Stores.inMemoryKeyValueStore(GENERELL_OPERATION_STORE),
        Serdes.String(),
        enhancedKafkaEventSerde
    )
    private val journalOperationSupplier = Stores.keyValueStoreBuilder(
        Stores.inMemoryKeyValueStore(JOURNALFOERING_OPERATION_STORE),
        Serdes.String(),
        enhancedKafkaEventSerde
    )
    private val oppgaveOperationSupplier = Stores.keyValueStoreBuilder(
        Stores.inMemoryKeyValueStore(OPPGAVE_OPERATION_STORE),
        Serdes.String(),
        enhancedKafkaEventSerde
    )

    val jfrTopologi: Topology
        get() {
            val streamsBuilder = StreamsBuilder()
            val serdeConfig: Map<String, String?> = kafkaSerdeConfig
            val valueGenericAvroSerde: Serde<GenericRecord> = GenericAvroSerde()
            valueGenericAvroSerde.configure(serdeConfig, false) // `false` for record values
            streamsBuilder.addStateStore(eventEnricherSupplier)
            streamsBuilder.addStateStore(generellOperationSupplier)
            streamsBuilder.addStateStore(journalOperationSupplier)
            streamsBuilder.addStateStore(oppgaveOperationSupplier)

            val aapenDokStream = streamsBuilder.stream(
                inputTopic, Consumed.with(Serdes.String(), valueGenericAvroSerde)
            )

            // Pre-processing
            val filteredEvents = filterGenerelleEvent(aapenDokStream)
            val kafkaEventKStream = convertGenericRecordToKafkaEvent(filteredEvents)

            val enrichedEventsStream = enrichKafkaEvent(kafkaEventKStream)

            val filterJournalpostToAuto = filterJournalpostToAuto(enrichedEventsStream)

            val processedGenerellStream = generellOperations(filterJournalpostToAuto)

            // Finish journalpost
            journalfoerJournalpost(processedGenerellStream)

            return streamsBuilder.build()
        }

    private fun filterGenerelleEvent(inputStream: KStream<String, GenericRecord>): KStream<String, GenericRecord> {
        return inputStream
            .filter { _, genericRecord ->
                genericRecord["temaNytt"].toString() == "SYK" && eventType.contains(
                    genericRecord["hendelsesType"].toString()
                )
            }
    }

    private fun convertGenericRecordToKafkaEvent(genericRecordString: KStream<String, GenericRecord>): KStream<String, KafkaEvent> {
        return genericRecordString.mapValues { genericRecord: GenericRecord ->
            objectMapper.readValue<KafkaEvent>(
                genericRecord.toString()
            )
        }
    }

    private fun enrichKafkaEvent(kafkaEventKStream: KStream<String, KafkaEvent>): KStream<String, EnrichedKafkaEvent> {
        val enrichedKafkaEvent = kafkaEventKStream.transform(eventEnricherTransformerSupplier, ENRICHER_OPERATION_STORE)
        val toManuell = enrichedKafkaEvent
            .filter { _, jfrEnhancedKafkaEvent -> jfrEnhancedKafkaEvent.isToManuell && !jfrEnhancedKafkaEvent.isToIgnore }

        sendToJfrManuellOppretter(toManuell)

        return enrichedKafkaEvent.filter { _, jfrEnhancedKafkaEvent -> !jfrEnhancedKafkaEvent.isToManuell && !jfrEnhancedKafkaEvent.isToIgnore }
    }

    private fun filterJournalpostToAuto(enrichedKafkaEventStream: KStream<String, EnrichedKafkaEvent>): KStream<String, EnrichedKafkaEvent> {
        val toAuto = enrichedKafkaEventStream
            .filter { _, jfrEnhancedKafkaEvent ->
                SkjemaMetadata.inAutoList(jfrEnhancedKafkaEvent.tema, jfrEnhancedKafkaEvent.skjema)
            }
            .peek { _, enrichedKafkaEvent ->
                logWithCorrelationId(
                    enrichedKafkaEvent,
                    "Forsøker automatisk behandling på journalpost ${enrichedKafkaEvent.journalpostId} med tema ${enrichedKafkaEvent.tema} og skjema ${enrichedKafkaEvent.skjema}"
                )
            }

        val toManuell = enrichedKafkaEventStream.filterNot { _, jfrEnhancedKafkaEvent ->
            SkjemaMetadata.inAutoList(jfrEnhancedKafkaEvent.tema, jfrEnhancedKafkaEvent.skjema)
        }

        sendToJfrManuellOppretter(toManuell)

        return toAuto
    }

    private fun generellOperations(enrichedKafkaEventStream: KStream<String, EnrichedKafkaEvent>): KStream<String, EnrichedKafkaEvent> {
        val afterGenerellOperationStream = enrichedKafkaEventStream
            .transform(generellOperationsTransformerSupplier, GENERELL_OPERATION_STORE)

        val toManuell = afterGenerellOperationStream
            .filter { _, enrichedKafkaEvent -> enrichedKafkaEvent.isToManuell }

        sendToJfrManuellOppretter(toManuell)

        return afterGenerellOperationStream.filterNot { _, enrichedKafkaEvent -> enrichedKafkaEvent.isToManuell }
    }

    private fun journalfoerJournalpost(oppgaveFilteredStream: KStream<String, EnrichedKafkaEvent>) {
        val journalpostAfterJournalfoering = oppgaveFilteredStream
            .transform(journalOperationsTransformerSupplier, JOURNALFOERING_OPERATION_STORE)

        val toManuell = journalpostAfterJournalfoering
            .filter { _, enrichedKafkaEvent -> enrichedKafkaEvent.isToManuell }

        journalpostAfterJournalfoering
            .filterNot { _, enrichedKafkaEvent -> enrichedKafkaEvent.isToManuell }
            .peek { _, enrichedKafkaEvent ->
                incJfrAutoProcess(
                    enrichedKafkaEvent!!
                )
            }

        sendToJfrManuellOppretter(toManuell)
    }

    private fun sendToJfrManuellOppretter(manuelle: KStream<String, EnrichedKafkaEvent>) {
        manuelle.foreach { _, enrichedKafkaEvent ->
            feilregistrer.feilregistrerOppgave(
                enrichedKafkaEvent!!
            )
        }
        manuelle.peek { _, enrichedKafkaEvent ->
            incJfrManuallProcess(
                enrichedKafkaEvent,
                SkjemaMetadata.inAutoList(enrichedKafkaEvent.tema, enrichedKafkaEvent.skjema)
            )
        }
        manuelle.peek { _, enrichedKafkaEvent ->
            logWithCorrelationId(
                enrichedKafkaEvent,
                "Journalposten: ${enrichedKafkaEvent.journalpostId} sendes til manuell-oppretter"
            )
        }
        manuelle.transform(oppgaveOperationsTransformerSupplier, OPPGAVE_OPERATION_STORE)
    }

    private fun logWithCorrelationId(enrichedKafkaEvent: EnrichedKafkaEvent, s: String) {
        MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
        log.info(s)
        MDC.clear()
    }

    companion object {
        private val log = LoggerFactory.getLogger(JfrTopologies::class.java)
        private val eventType = listOf("MidlertidigJournalført", "Mottatt", "JournalpostMottatt")
        private const val ENRICHER_OPERATION_STORE = "enrichoperations"
        private const val GENERELL_OPERATION_STORE = "generelloperations"
        private const val JOURNALFOERING_OPERATION_STORE = "journalfoeringoperations"
        private const val OPPGAVE_OPERATION_STORE = "oppgave"
    }
}
